lazy_loader/__init__.py

```diff
@@ -160,38 +166,72 @@ def myfunc():
         Actual loading of the module occurs upon first attribute request.
 
     """
-    try:
-        return sys.modules[fullname]
-    except KeyError:
-        pass
+    module = sys.modules.get(fullname)
+    have_module = module is not None
+
+    # Most common, short-circuit
+    if have_module and require is None:
+        return module
+
+    spec = None
+    if not have_module:
+        spec = importlib.util.find_spec(fullname)
+        have_module = spec is not None
+
+    if not have_module:
+        not_found_message = f"No module named '{fullname}'"
+    elif require is not None:
+        # Old style lazy loading to avoid polluting sys.modules
+        import packaging.requirements
+
+        req = packaging.requirements.Requirement(require)
+        try:
+            have_module = req.specifier.contains(
+                importlib_metadata.version(req.name),
+                prereleases=True,
+            )
+        except importlib_metadata.PackageNotFoundError as e:
+            raise ValueError(
+                f"Found module '{fullname}' but cannot test requirement '{require}'. "
+                "Requirements must match distribution name, not module name."
```

Comment:
What does this refer to?

Reply:
I suspect a common error will be to do something like `yaml = load('yaml', 'yaml >= 5')`. The `yaml` module is provided by the `pyyaml` distribution, so it needs to be `yaml = load('yaml', 'pyyaml >= 5')`.

This is necessary because not all modules provide a `__version__`, but all distributions must provide version metadata, and that's what's retrieved by `importlib.metadata.version`.